ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,244 @@
+      final Expression valueExp = insertValues.getValues().get(i);
+
+      values.put(column, new ExpressionResolver(columnSchema, column).process(valueExp, null));
+    }
+
+    if (keyField.isPresent()) {
+      final String key = keyField.get();
+      final Object keyValue = values.get(key);
+      final Object rowKeyValue = values.get(SchemaUtil.ROWKEY_NAME);
+
+      if ((keyValue != null) ^ (rowKeyValue != null)) {
+        values.putIfAbsent(key, rowKeyValue);
+        values.putIfAbsent(SchemaUtil.ROWKEY_NAME, keyValue);
+      } else if (!Objects.equals(keyValue, rowKeyValue)) {
+        throw new KsqlException(
+            String.format(
+                "Expected ROWKEY and %s to match but got %s and %s respectively.",
+                key, rowKeyValue, keyValue));
+      }
+    }
+
+    values.putIfAbsent(SchemaUtil.ROWTIME_NAME, clock.getAsLong());
+
+    return new GenericRow(
+        dataSource.getSchema()
+            .fields()
+            .stream()
+            .map(Field::name)
+            .map(values::get)
+            .collect(Collectors.toList())
+    );
+  }
+
+  @SuppressWarnings("unchecked") // we know that key is String
+  private byte[] serializeKey(final Object keyValue, final StructuredDataSource<?> dataSource) {
+    if (keyValue == null) {
+      return null;
+    }
+
+    return ((Serde<String>) dataSource.getKeySerdeFactory().create())
+        .serializer()
+        .serialize(dataSource.getKafkaTopicName(), keyValue.toString());
+  }
+
+  private byte[] serializeRow(
+      final GenericRow row,
+      final StructuredDataSource<?> dataSource,
+      final KsqlConfig config,
+      final ServiceContext serviceContext
+  ) {
+    final Serde<GenericRow> rowSerde = dataSource.getKsqlTopicSerde()
+        .getGenericRowSerde(
+            dataSource.getSchema(),
+            config,
+            serviceContext.getSchemaRegistryClientFactory(),
+            "insertValues",
+            ProcessingLogContext.create());
+    return rowSerde.serializer().serialize(dataSource.getKafkaTopicName(), row);
+  }
+
+  private static class ExpressionResolver extends AstVisitor<Object, Void> {
+
+    private static final Map<Type, Function<Number, Object>> CASTER =
+        ImmutableMap.<Type, Function<Number, Object>>builder()
+            .put(Type.INT8, Number::shortValue)
+            .put(Type.INT32, Number::intValue)
+            .put(Type.INT64, Number::longValue)
+            .put(Type.FLOAT32, Number::floatValue)
+            .put(Type.FLOAT64, Number::doubleValue)
+            .build();
+
+    private final Schema schema;
+    private final String field;
+
+    ExpressionResolver(final Schema schema, final String field) {
+      this.schema = schema;
+      this.field = field;
+    }
+
+    @Override
+    protected Object visitLiteral(final Literal node, final Void context) {
+      final Object value = node.getValue();
+      if (node instanceof NullLiteral || value == null) {
+        if (schema.isOptional()) {
```

Comment:
We should do this same verification for omitted columns. In practice, I don't think we actually have non-optional schemas so not a huge deal.